core/Services/Calendar/Store.vala

```diff
@@ -1,307 +1,658 @@
 
-            // Add sources
-            registry.list_sources (E.SOURCE_EXTENSION_CALENDAR).foreach ((source) => {
-                E.SourceCalendar cal = (E.SourceCalendar)source.get_extension (E.SOURCE_EXTENSION_CALENDAR);
-                if (cal.selected == true && source.enabled == true) {
-                    add_source (source);
-                }
+            registry.source_added.connect (registry_source_added);
+            registry.source_changed.connect (registry_source_changed);
+            registry.source_removed.connect (registry_source_removed);
+
+            // Connect to Sources
+            sources_list ().foreach ((source) => {
+                registry_source_added (source);
             });
-        } catch (GLib.Error error) {
+
+        } catch (Error error) {
             critical (error.message);
         }
     }
 
-    //--- Public Methods ---//
+    //--- Public Source API ---//
+
+    public void source_add (E.Source source) {
+        debug ("Adding source '%s'", source.dup_display_name ());
+
+        sources_add_transaction.add (source);
+        source_added (source);
+
+        var sources = new List<E.Source> ();
+        sources.append (source);
+
+        registry.create_sources.begin (sources, null, (obj, res) => {
+            Idle.add (() => {
+                try {
+                    registry.create_sources.end (res);
+                } catch (Error e) {
+                    foreach (var transactional_source in sources) {
+                        sources_add_transaction.remove (source);
+                        source_removed (source);
+                    }
+                    error_received (e);
+                    critical (e.message);
+                }
+                return Source.REMOVE;
+            });
+        });
+    }
 
-    public void add_event (E.Source source, ECal.Component event) {
-        add_event_async.begin (source, event);
+    public void source_connect (E.Source source) {
+        source_connect_async.begin (source);
     }
 
-    public bool calclient_is_readonly (E.Source source) {
+    public void source_remove (E.Source source) {
+        debug ("Removing source '%s'", source.dup_display_name ());
+
+        source_removed (source);
+        source.remove.begin (null, (obj, res) => {
+            Idle.add (() => {
+                try {
+                    source.remove.end (res);
+                } catch (Error e) {
+                    source_added (source);
+                    error_received (e);
+                    critical (e.message);
+                }
+                return Source.REMOVE;
+            });
+        });
+    }
+
+    public void source_disconnect (E.Source source) {
+        source_disconnect_async.begin (source);
+    }
+
+    public E.Source source_get_with_uid (string uid) {
+        return registry.ref_source (uid);
+    }
+
+    public string? source_get_ancestor_display_name (E.Source source) {
```

Comment:
This function is never used, at least in this state.

Reply:
True that. This one [is used in the Tasks app](https://github.com/elementary/tasks/blob/2226b036bc8870e7a5944a411934bbf2cea67f58/src/MainWindow.vala#L226). I guess we should decide whether we want to use `source_get_ancestor_display_name` or `source_get_location` as for the section header names.

From what I can tell, these are two different approaches and I didn't want to mix them up - so I implemented both methods to not change behaviour in the Calendar app.